### PR TITLE
feat: add name to `envd envs describe`

### PR DESCRIPTION
**Pull request by AlexXi19**
Signed-off-by: AlexXi19 <alex2001314jjj@gmail.com>

https://github.com/tensorchord/envd/issues/551

Show service name in ports list as specified by the issue.  

### Testing 
Inspected `envd envs describe` for both jupyter notebook and R studio and verified that the names are correct. 
image

image

### CODE REVIEW

**Review by kemingy — CHANGES_REQUESTED** [review]

File: pkg/app/env_describe.go [hunk 1]
```diff
@@ -70,16 +74,43 @@ func getEnvironmentDescriptions(clicontext *cli.Context) error {
 		return errors.Wrap(err, "failed to list dependencies")
 	}
 
+	images, err := envdEngine.ListImage(clicontext.Context)
+	if err != nil {
+		return errors.Wrap(err, "Failed to list image")
+	}
```
[the remark this answers is not in the document]
Comment (gaocegege) [reply pkg/app/env_describe.go:80]: Same question here.

[the remark this answers is not in the document]
Comment (AlexXi19) [reply pkg/app/env_describe.go:80]: I was browsing around for the label data and I found it [here](https://github.com/tensorchord/envd/blob/e04be4836206c4a278414b84677101101b97f1a0/pkg/types/envd.go#L87) with the `EnvdImage` type from the `ImageClient`, is there another way to get the label information from the environment?

File: pkg/types/envd.go [hunk 2]
```diff
@@ -152,6 +152,7 @@ type RepoInfo struct {
 }
 
 type PortBinding struct {
+	Name     string
```
[the remark this answers is not in the document]
Comment (AlexXi19) [reply pkg/types/envd.go:155]: The name field is added to the struct later after the function is returned, maybe this is a bad practice? Something I would like to get your opinion on is should I modify the struct and have `envdEngine.ListEnvPortBinding` return a `PortBinding` struct with the `Name` field (doing label fetching logic in this function) or should I not have the Name in the struct and add the name outside of this function?

File: pkg/app/env_describe.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexXi19) [reply pkg/app/env_describe.go:80]: Maybe I should make another function in the `EnvironmentClient` to do this or add it to the `ListEnvPortBindings` function.

[the remark this answers is not in the document]
Comment (gaocegege) [reply pkg/app/env_describe.go:80]: Yep, I think so. You can add it to ListEnvPortBinding

File: pkg/types/envd.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (gaocegege) [reply pkg/types/envd.go:155]: I think we can set the name in the func `ListEnvPortBinding`

[the remark this answers is not in the document]
Comment (AlexXi19) [reply pkg/types/envd.go:155]: OK, the logic will be the same, just moved to `ListEnvPortBinding` right?

**Comment by AlexXi19** [comment]
Addressed the comments and moved label name fetching logic to `ListEnvPortBinding`

File: pkg/envd/docker.go [hunk 3]
```diff
@@ -193,6 +194,31 @@ func (e dockerEngine) ListEnvPortBinding(ctx context.Context, env string) ([]typ
...
+			for _, environmentPort := range environmentPorts {
+				portMap[fmt.Sprint(environmentPort.Port)] = environmentPort.Name
+			}
+		}
+	}
+
+	for index, label := range ports {
+		label.Name = portMap[label.Port]
+		ports[index] = label
+	}
```
Comment (kemingy) [inline pkg/envd/docker.go:220]: Can you get the same information from `ctr, err := e.ContainerInspect(ctx, env)`? If so, I think you should add some logic to `NewPortBindingFromContainerJSON`.

BTW, if you modify the `dockerEngine.ListEnvPortBinding`, you should also modify `envdServerEngine.ListEnvPortBinding`.

Comment (AlexXi19) [reply pkg/envd/docker.go:220]: I remember looking into `ContainerInspect` and not finding any label/name info, I will look again. I will also modify `envdServerEngine.ListEnv...`

Comment (AlexXi19) [reply pkg/envd/docker.go]: I found label info in `ctr` 😓 and moved the logic into `NewPortBindingFromContainerJSON`

File: pkg/envd/envdserver.go [hunk 4]
```diff
@@ -219,6 +219,7 @@ func (e *envdServerEngine) ListEnvPortBinding(
 	// TODO(gaocegege): Remove hard coded.
 	res := []types.PortBinding{
 		{
+			Name:     "ssh",
```
[the remark this answers is not in the document]
Comment (gaocegege) [reply pkg/envd/envdserver.go:222]: Currently, we still use 2222 as a hard-coded port. Let's open an issue in envd-server to keep track of it.

[the remark this answers is not in the document]
Comment (gaocegege) [reply pkg/envd/envdserver.go:222]: https://github.com/tensorchord/envd-server/issues/76

[the remark this answers is not in the document]
Comment (AlexXi19) [reply pkg/envd/envdserver.go:222]: Yes. I matched what's provided by the labels, which is `ssh`. Let me know if you want me to change the labels for the docker engine [here](https://github.com/tensorchord/envd/blob/41272d97df5cad530a006a73da246eb82e21fe22/pkg/lang/ir/compile.go#L160). 

For the hardcode, I'm also happy to take a look at envdserver and replace it with the real implementation.

[the remark this answers is not in the document]
Comment (gaocegege) [reply pkg/envd/envdserver.go:222]: The change may include the envd context implementation and envd-server. You can give it a try if you are interested!

[the remark this answers is not in the document]
Comment (AlexXi19) [reply pkg/envd/envdserver.go:222]: Ok, maybe later haha, for now should I just change `ssh` => `sshd`?

[the remark this answers is not in the document]
Comment (kemingy) [reply pkg/envd/envdserver.go:222]: > Yes. I matched what's provided by the labels, which is `ssh`. Let me know if you want me to change the labels for the docker engine [here](https://github.com/tensorchord/envd/blob/41272d97df5cad530a006a73da246eb82e21fe22/pkg/lang/ir/compile.go#L160).

Yep.

> For the hardcode, I'm also happy to take a look at envdserver and replace it with the real implementation.

I think it's doable now. `EnvironmentGet` will return the image info which contains the image labels.

**Pull request merged**